lua/apisix/plugins/batch-processor.lua

```diff
@@ -0,0 +1,179 @@
+
+    if premature then
+        return
+    end
+
+    if now() - batch_processor.last_entry_t >= batch_processor.inactive_timeout or
+            now() - batch_processor.first_entry_t >= batch_processor.buffer_duration then
+        ngx_log(DEBUG, fmt("BatchProcessor[%s] buffer duration exceeded, activating buffer flush", batch_processor.name))
+        batch_processor:process_buffer()
+        batch_processor.isTimerRunning = false
+        return
+    end
+
+    -- buffer duration did not exceed or the buffer is active, extending the timer
+    ngx_log(DEBUG, fmt("BatchProcessor[%s] extending buffer timer", batch_processor.name))
+    create_buffer_timer(batch_processor)
+end
+
+
+create_buffer_timer = function(batch_processor)
+    local hdl, err = timer_at(batch_processor.inactive_timeout, flush_buffer, batch_processor)
+    if not hdl then
+        core.log.error("failed to create buffer timer: ", err)
+        return
+    end
+    batch_processor.isTimerRunning = true
+end
+
+
+function Batch_Processor:new(func, config)
+    local ok, err = core.schema.check(schema, config)
+
+    if not ok then
+        return err
+    end
+
+    if not(type(func) == "function") then
+        return nil, "Invalid argument, arg #1 must be a function"
+    end
+
+    local batch_processor = {
+        func = func,
+        buffer_duration = config.buffer_duration,
+        inactive_timeout = config.inactive_timeout,
+        max_retry_count = config.max_retry_count,
+        batch_max_size = config.batch_max_size,
+        retry_delay = config.retry_delay,
+        name = config.name,
+        batch_to_process = {},
+        entry_buffer = { entries = {}, retry_count = 0},
+        isTimerRunning = false,
+        first_entry_t = 0,
+        last_entry_t = 0
+    }
+
+    return setmetatable(batch_processor, Batch_Processor_mt)
+end
+
+
+function Batch_Processor:push(entry)
+    -- if the batch size is one then immediately send for processing
+    if self.batch_max_size == 1 then
+        local batch = { entries = { entry }, retry_count = 0 }
+        schedule_func_exec(self, 0, batch)
+    end
+
+    local entries = self.entry_buffer.entries
+    table.insert(entries, entry)
+
+    if #entries == 1 then
+        self.first_entry_t = now()
+    end
+    self.last_entry_t = now()
+
+    if self.batch_max_size <= #entries then
+        ngx_log(DEBUG, fmt("batch processor[%s] batch max size has exceeded", self.name))
+        self:process_buffer()
+    elseif not self.isTimerRunning then
+        create_buffer_timer(self)
+    end
+
+end
+
+
+function Batch_Processor:process_buffer()
+    if #self.entry_buffer.entries > 0 then
+        ngx_log(DEBUG, fmt("tranferring buffer entries to processing pipe line, buffercount[%d]", #self.entry_buffer.entries))
+        self.batch_to_process[#self.batch_to_process + 1] = self.entry_buffer
+        self.entry_buffer = { entries = {}, retry_count = 0 }
+    end
+
+    if(#self.batch_to_process > 0) then
+        repeat
```

Comment:
Avoid using `table.move`, this API is slow.

Reply:
Hi @membphis can I get some guidance on this? To avoid using the remove method, 
1. Should I use a deep copy approach 
```lua
local temp = deepcopy(self.batch_to_process)
self.batch_to_process = {}
-- start processing the temp batch
```

2. Or use a [reindexing approach](https://stackoverflow.com/a/53038524/6164611) as discussed in this answer?

Reply:
How about this style?

```lua
for _, batch in ipairs(self.batch_to_process) do
    schedule_func_exec(self, 0, batch)
end
```

Reply:
Thanks, refactored.